File: ios/KibleBulucu/CompassView.swift
//
//  CompassView.swift
//  KibleBulucu
//
//  Created by Yunus Elçi on 27.05.2025.
//

import SwiftUI

struct CompassView: View {
    let currentHeading: Double
    let qiblaDirection: Double
    
    // State to track the smooth rotation for animation
    @State private var smoothRotation: Double = 0
    
    private var targetRotation: Double {
        let rotation = qiblaDirection - currentHeading
        return normalizeAngle(rotation)
    }
    
    // Check if phone direction is aligned with qibla direction
    private var isAligned: Bool {
        let angleDifference = abs(normalizeAngle(targetRotation))
        return angleDifference <= 3 // Allow 3 degree tolerance
    }
    
    // Normalize angle to be between -180 and 180
    private func normalizeAngle(_ angle: Double) -> Double {
        var normalizedAngle = angle.truncatingRemainder(dividingBy: 360)
        if normalizedAngle > 180 {
            normalizedAngle -= 360
        } else if normalizedAngle < -180 {
            normalizedAngle += 360
        }
        return normalizedAngle
    }
    
    // Calculate the shortest angular distance between two angles
    private func shortestAngularDistance(from: Double, to: Double) -> Double {
        let difference = to - from
        let normalizedDifference = normalizeAngle(difference)
        return normalizedDifference
    }
    
    // Update smooth rotation to avoid large jumps
    private func updateSmoothRotation() {
        let distance = shortestAngularDistance(from: smoothRotation, to: targetRotation)
        smoothRotation += distance
    }

    var body: some View {
        VStack(spacing: 30) {
            // Main rotating arrow
            Image(systemName: "arrow.up.circle.fill")
                .font(.system(size: 120))
                .foregroundColor(isAligned ? .green : .red)
                .rotationEffect(.degrees(smoothRotation))
                .animation(.easeInOut(duration: 0.3), value: smoothRotation)
                .animation(.easeInOut(duration: 0.3), value: isAligned)
                .shadow(color: .black.opacity(0.1), radius: 4, x: 0, y: 2)
                .onChange(of: targetRotation) { _ in
                    updateSmoothRotation()
                }
                .onAppear {
                    // Initialize smooth rotation on first appearance
                    smoothRotation = targetRotation
                }
            
            // Direction info
            VStack(spacing: 8) {
                Text("KIBLE")
                    .font(.caption)
                    .fontWeight(.medium)
                    .foregroundColor(.secondary)
                    .tracking(2)
                
                Text("\(Int(qiblaDirection.rounded()))°")
                    .font(.title2)
                    .fontWeight(.semibold)
                    .foregroundColor(.primary)
                
                // Alignment status
                Text(isAligned ? "HİZALANDI" : "HİZALA")
                    .font(.caption)
                    .fontWeight(.bold)
                    .foregroundColor(isAligned ? .green : .red)
                    .animation(.easeInOut(duration: 0.3), value: isAligned)
            }
        }
    }
}

#Preview {
    CompassView(currentHeading: 45, qiblaDirection: 136)
        .padding()
} 

File: app/src/main/java/com/yunuselci/KibleBulucu/CompassView.kt
package com.yunuselci.KibleBulucu

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.abs
import kotlin.math.roundToInt

@Composable
fun CompassView(
    currentHeading: Double,
    qiblaDirection: Double,
    modifier: Modifier = Modifier
) {
    // State to track the smooth rotation for animation
    var smoothRotation by remember { mutableDoubleStateOf(0.0) }
    
    val targetRotation = remember(qiblaDirection, currentHeading) {
        val rotation = qiblaDirection - currentHeading
        normalizeAngle(rotation)
    }
    
    // Check if phone direction is aligned with qibla direction
    val isAligned = remember(targetRotation) {
        val angleDifference = abs(normalizeAngle(targetRotation))
        angleDifference <= 3.0 // Allow 3 degree tolerance
    }
    
    // Update smooth rotation to avoid large jumps
    LaunchedEffect(targetRotation) {
        val distance = shortestAngularDistance(smoothRotation, targetRotation)
        smoothRotation += distance
    }
    
    // Initialize smooth rotation on first appearance
    LaunchedEffect(Unit) {
        smoothRotation = targetRotation
    }
    
    // Animated values
    val animatedRotation by animateFloatAsState(
        targetValue = smoothRotation.toFloat(),
        animationSpec = tween(durationMillis = 150),
        label = "rotation"
    )
    
    val arrowColor by animateColorAsState(
        targetValue = if (isAligned) Color(0xFF4CAF50) else Color(0xFFF44336),
        animationSpec = tween(durationMillis = 200),
        label = "color"
    )
    
    val statusColor by animateColorAsState(
        targetValue = if (isAligned) Color(0xFF4CAF50) else Color(0xFFF44336),
        animationSpec = tween(durationMillis = 200),
        label = "statusColor"
    )
    
    // Theme-aware shadow color for better blending in light mode
    val shadowColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.08f)
    
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(30.dp)
    ) {
        // Main rotating arrow with filled circle background (like iOS)
        Box(
            modifier = Modifier
                .size(120.dp)
                .rotate(animatedRotation)
                .shadow(
                    elevation = 1.dp,
                    spotColor = shadowColor,
                    ambientColor = shadowColor,
                    shape = CircleShape
                )
                .clip(CircleShape)
                .background(arrowColor),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = Icons.Filled.ArrowUpward,
                contentDescription = "Qibla Direction Arrow",
                modifier = Modifier.size(60.dp),
                tint = Color.White
            )
        }
        
        // Direction info
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "KIBLE",
                fontSize = 12.sp,
                fontWeight = FontWeight.Medium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                letterSpacing = 2.sp
            )
            
            Text(
                text = "${qiblaDirection.roundToInt()}°",
                fontSize = 20.sp,
                fontWeight = FontWeight.SemiBold,
                color = MaterialTheme.colorScheme.onSurface
            )
            
            // Alignment status
            Text(
                text = if (isAligned) "HİZALANDI" else "HİZALA",
                fontSize = 12.sp,
                fontWeight = FontWeight.Bold,
                color = statusColor,
                modifier = Modifier.padding(top = 8.dp)
            )
        }
    }
}

// Normalize angle to be between -180 and 180
private fun normalizeAngle(angle: Double): Double {
    var normalizedAngle = angle % 360
    if (normalizedAngle > 180) {
        normalizedAngle -= 360
    } else if (normalizedAngle < -180) {
        normalizedAngle += 360
    }
    return normalizedAngle
}

// Calculate the shortest angular distance between two angles
private fun shortestAngularDistance(from: Double, to: Double): Double {
    val difference = to - from
    return normalizeAngle(difference)
} 